src/sentry/middleware/ratelimit.py

```diff
@@ -15,80 +15,92 @@
-            # TODO: put these fields into their own object
-            request.will_be_rate_limited = False
-            request.rate_limit_category = None
-            request.rate_limit_uid = uuid.uuid4().hex
-            request.rate_limit_key = get_rate_limit_key(view_func, request)
-            if request.rate_limit_key is None:
-                return
-            category_str = request.rate_limit_key.split(":", 1)[0]
-            request.rate_limit_category = category_str
+            rate_limit_uid = uuid.uuid4().hex
+            view_func = resolve(request.path).func
+            # in deprecated django middleware the view function is passed in to process_view
+            # but we can still access it this way through the request
+            rate_limit_key = get_rate_limit_key(view_func, request)
+            if rate_limit_key is None:
+                return self.get_response(request)
 
+            category_str = rate_limit_key.split(":", 1)[0]
             rate_limit = get_rate_limit_value(
                 http_method=request.method,
                 endpoint=view_func.view_class,
                 category=RateLimitCategory(category_str),
             )
             if rate_limit is None:
-                return
+                return self.get_response(request)
 
-            request.rate_limit_metadata = above_rate_limit_check(
-                request.rate_limit_key, rate_limit, request.rate_limit_uid
+            rate_limit_metadata = above_rate_limit_check(
+                rate_limit_key, rate_limit, rate_limit_uid, category_str
             )
             # TODO: also limit by concurrent window once we have the data
             rate_limit_cond = (
-                request.rate_limit_metadata.rate_limit_type != RateLimitType.NOT_LIMITED
+                rate_limit_metadata.rate_limit_type != RateLimitType.NOT_LIMITED
                 if ENFORCE_CONCURRENT_RATE_LIMITS
-                else request.rate_limit_metadata.rate_limit_type == RateLimitType.FIXED_WINDOW
+                else rate_limit_metadata.rate_limit_type == RateLimitType.FIXED_WINDOW
             )
             if rate_limit_cond:
-                request.will_be_rate_limited = True
                 enforce_rate_limit = getattr(view_func.view_class, "enforce_rate_limit", False)
                 if enforce_rate_limit:
-                    return HttpResponse(
+                    response = HttpResponse(
                         {
                             "detail": DEFAULT_ERROR_MESSAGE.format(
-                                limit=request.rate_limit_metadata.limit,
-                                window=request.rate_limit_metadata.window,
+                                limit=rate_limit_metadata.limit,
+                                window=rate_limit_metadata.window,
                             )
                         },
                         status=429,
                     )
+                    self.add_headers(response, rate_limit_metadata)
```

Comment:
You have to remember to call `self.add_headers` twice to make this code work properly. Can it be reorganized to only be called once so that people modifying this code in the future don't have to remember to?